src/core/tests/Test_AptitudePackageManager.py

```diff
@@ -349,6 +367,123 @@ def test_update_image_default_patch_mode_raises_exception(self):
+
+        runtime.package_manager._AptitudePackageManager__pro_client_prereq_met = backup_AptitudePackageManager__pro_client_prereq_met
+        version_mock.mock_unimport_uaclient_version_module()
+        reboot_mock.mock_unimport_uaclient_reboot_required_module()
+
+    def test_is_reboot_pending_test_mismatch(self):
+        version_mock = MockVersionResult()
+        version_mock.mock_import_uaclient_version_module('version', 'mock_version')
+        reboot_mock = MockRebootRequiredResult()
+        reboot_mock.mock_import_uaclient_reboot_required_module('reboot_required', 'mock_reboot_required_return_yes')
+        runtime = RuntimeCompositor(ArgumentComposer().get_composed_arguments(), True, Constants.APT)
+        backup_package_manager_is_reboot_pending = runtime.package_manager.ubuntu_pro_client.is_reboot_pending
+        runtime.package_manager.ubuntu_pro_client.is_reboot_pending = self.mock_is_reboot_pending_returns_True
+        runtime.package_manager._AptitudePackageManager__pro_client_prereq_met = True
+
+        # test should return true as we fall back to Ubuntu Pro Client api`s result.
+        self.assertTrue(runtime.package_manager.is_reboot_pending())
+
+        runtime.package_manager.ubuntu_pro_client.is_reboot_pending = backup_package_manager_is_reboot_pending
+        version_mock.mock_unimport_uaclient_version_module()
+        reboot_mock.mock_unimport_uaclient_reboot_required_module()
+
+    def test_is_reboot_pending_test_raises_exception(self):
+        version_mock = MockVersionResult()
+        version_mock.mock_import_uaclient_version_module('version', 'mock_version')
+        reboot_mock = MockRebootRequiredResult()
+        reboot_mock.mock_import_uaclient_reboot_required_module('reboot_required', 'mock_reboot_required_return_yes')
+        runtime = RuntimeCompositor(ArgumentComposer().get_composed_arguments(), True, Constants.APT)
+        backup_package_manager_do_processes_require_restart = runtime.package_manager.do_processes_require_restart
+        runtime.package_manager.do_processes_require_restart = self.mock_do_processes_require_restart_raises_exception
+        backup_package_manager_is_reboot_pending = runtime.package_manager.ubuntu_pro_client.is_reboot_pending
+        runtime.package_manager.ubuntu_pro_client.is_reboot_pending = self.mock_is_reboot_pending_returns_False
+        runtime.package_manager._AptitudePackageManager__pro_client_prereq_met = True
+
+        # test returns true because, we return True if there is exception.
+        self.assertTrue(runtime.package_manager.is_reboot_pending())
+
+        runtime.package_manager.do_processes_require_restart = backup_package_manager_do_processes_require_restart
+        runtime.package_manager.ubuntu_pro_client.is_reboot_pending = backup_package_manager_is_reboot_pending
+        version_mock.mock_unimport_uaclient_version_module()
+        reboot_mock.mock_unimport_uaclient_reboot_required_module()
```

Comment:
These and the mock_import calls are used in multiple functions, can be extracted to one mock_import<>() and one mock_unimport() function, which are then called in tests

Reply:
To add to this, see if there can be a setup_is_reboot_pending_test() function that extracts out common setup portions between these tests (eg: mock_imports, mocking is_reboot_pending, setting _AptitudePackageManager__pro_client_prereq_met, etc)

Reply:
The mock imports can have different parameters based on the tests which will make it difficult to to refactor into a common setup code . Whatever was common between these modules are already refactored into a common class (MockSystemModules class in Test_UbuntuProClient.py) which i am planning to use for updates module too.